rust/kernel/hwrng.rs

```diff
@@ -0,0 +1,223 @@
+//! Hardware Random Number Generator.
+//!
+//! C header: [`include/linux/hw_random.h`](../../../../include/linux/hw_random.h)
+
+use alloc::{boxed::Box, slice::from_raw_parts_mut};
+
+use crate::{
+    bindings, c_types, error::from_kernel_result, str::CStr, to_result, types::PointerWrapper,
+    Error, Result, ScopeGuard,
+};
+
+use core::{cell::UnsafeCell, marker::PhantomData, pin::Pin};
+
+/// This trait is implemented in order to provide callbacks to `struct hwrng`.
+pub trait Operations: Sized + 'static {
+    /// The methods to use to populate [`struct hwrng`].
+    const TO_USE: ToUse;
+
+    /// The pointer type that will be used to hold user-defined data type.
+    type Data: PointerWrapper = Box<Self>;
+
+    /// Initialization callback, can be left undefined.
+    fn init(_data: <Self::Data as PointerWrapper>::Borrowed<'_>) -> Result {
+        Err(Error::EINVAL)
+    }
+
+    /// Cleanup callback, can be left undefined.
+    fn cleanup(_data: <Self::Data as PointerWrapper>::Borrowed<'_>) {}
+
+    /// Read data into the provided buffer.
+    /// Drivers can fill up to max bytes of data into the buffer.
+    /// The buffer is aligned for any type and its size is a multiple of 4 and >= 32 bytes.
+    fn read(
+        data: <Self::Data as PointerWrapper>::Borrowed<'_>,
+        buffer: &mut [i8],
+        wait: bool,
+    ) -> Result<u32>;
+}
+
+/// Registration structure for Hardware Random Number Generator driver.
+pub struct Registration<T: Operations> {
+    hwrng: UnsafeCell<bindings::hwrng>,
+    registered: bool,
+    _p: PhantomData<T>,
+}
+
+impl<T: Operations> Registration<T> {
+    /// Creates new instance of registration.
+    ///
+    /// The data must be registered.
+    pub fn new() -> Self {
+        Self {
+            hwrng: UnsafeCell::new(bindings::hwrng::default()),
+            registered: false,
+            _p: PhantomData::default(),
+        }
+    }
+
+    /// Returns a registered and pinned, heap-allocated representation of the registration.
+    pub fn new_pinned(name: &CStr, quality: u16, data: T::Data) -> Result<Pin<Box<Self>>> {
+        let mut reg = Pin::from(Box::try_new(Self::new())?);
+        reg.as_mut().register(name, quality, data)?;
+        Ok(reg)
+    }
+
+    /// Registers a hwrng device within the rest of the kernel.
+    ///
+    /// It must be pinned because the memory block that represents
+    /// the registration may be self-referential.
+    pub fn register(self: Pin<&mut Self>, name: &CStr, quality: u16, data: T::Data) -> Result {
+        // SAFETY: We never move out of `this`.
+        let this = unsafe { self.get_unchecked_mut() };
+
+        if this.registered {
+            return Err(Error::EINVAL);
+        }
+
+        let data_pointer = <T::Data as PointerWrapper>::into_pointer(data);
+
+        // The registration is left uninitialized when registration fails to indicate failure.
+        let mut guard = ScopeGuard::new_with_data(this, |this| {
```

Comment:
I've just understood how `into_pointer()` is working. Yeah, this is the problem.

Shouldn't this method be unsafe then? This is exactly the place which may introduce
a potential leak and there is no even a single comment pointing about it.

Reply:
I have just understood how `into_pointer()` is working. I see the issue now.

Shouldn't this method be unsafe, then? This is exactly the place that may introduce a potential leak,
and there is not even a single comment pointing about it.

Reply:
> Shouldn't this method be unsafe, then? This is exactly the place that may introduce a potential leak,
> and there is not even a single comment pointing about it.

It shouldn't be unsafe because leaking is not considered unsafe in Rust: see the explicit item in the [reference](https://doc.rust-lang.org/reference/behavior-not-considered-unsafe.html). The simplest example is [forget](https://doc.rust-lang.org/core/mem/fn.forget.html) (which is not unsafe).

As far as improving documentation with comments warning about leaks, I'm all for improving that.

Reply:
> It shouldn't be unsafe because leaking is not considered unsafe in Rust: see the explicit item in the [reference](https://doc.rust-lang.org/reference/behavior-not-considered-unsafe.html). The simplest example is [forget](https://doc.rust-lang.org/core/mem/fn.forget.html) (which is not unsafe).

Thanks for clarifying that. This fact clears the confusion.
> 
> As far as improving documentation with comments warning about leaks, I'm all for improving that.

Okay, I'll create then an appropriate meta-issue for that topic soon.